lib/api/src/externals/function.rs

```diff
@@ -621,11 +621,39 @@ mod inner {
         };
     }
 
+    macro_rules! from_to_native_wasm_type_same_size {
+        ( $( $type:ty => $native_type:ty ),* ) => {
+            $(
+                #[allow(clippy::use_self)]
+                unsafe impl FromToNativeWasmType for $type {
+                    type Native = $native_type;
+
+                    #[inline]
+                    fn from_native(native: Self::Native) -> Self {
+                        unsafe {
+                            std::mem::transmute::<$native_type, $type>(native)
```

Comment:
Does this have to be an unsafe transmute? What does it mean if the endianness of the system is different, do we want to produce a different result? Could this be written with a from_ne_bytes / to_ne_bytes pair?

Reply:
I thought this wasn't possible, but it seems it is possible to do that since Rust 1.40.0 ; yeah we can do this without unsafe, I don't know if there's any overhead for doing so... I can check

Reply:
Interestingly, it looks like the `X::from_ne_bytes(Y::to_ne_bytes(value))` is the _most_ efficient way to express this, it gets compiled out entirely in 1.45.2 and 1.46.0, whereas the transmute doesn't in the test I just made

Reply:
I think it'd be reasonable to file a bug on rustc that transmute doesn't get compiled out in this case. Regardless we should prefer code that is safe and expressive, and if that just happens to get us better code generation, all the better!

Reply:
```Rust
use std::convert::TryInto;

#[inline(never)]
pub fn convert_try(i: u32) -> i32 {
    i.try_into().unwrap()
}

#[inline(never)]
pub fn convert_trans(i: u32) -> i32 {
    unsafe { std::mem::transmute(i) }
}

#[inline(never)]
pub fn convert_to_from(i: u32) -> i32 {
   i32::from_ne_bytes(u32::to_ne_bytes(i))
}
```

is the thing I tested, it's likely that the others would get compiled out if they're actually used, but only the bottom one was completely missing from the generated code.

with this being the generated code

```asm
example::convert_try:
        push    rax
        test    edi, edi
        js      .LBB1_1
        mov     eax, edi
        pop     rcx
        ret
.LBB1_1:
        lea     rdi, [rip + .L__unnamed_1]
        lea     rcx, [rip + .L__unnamed_2]
        lea     r8, [rip + .L__unnamed_3]
        mov     rdx, rsp
        mov     esi, 43
        call    qword ptr [rip + core::result::unwrap_failed@GOTPCREL]
        ud2

example::convert_trans:
        mov     eax, edi
        ret
```


Normally, I'd agree in regards to safe/expressive but this is potentially very, very hot code in Wasmer as it happens on every parameter every time the host calls into the guest, so I think it's worth us going out of our way to make sure things here are fast.